source/Irrlicht/COpenGLDriver.cpp

```diff
@@ -2255,6 +2255,209 @@ void COpenGLDriver::SAuxContext::setActiveDescriptorSet(uint32_t index, const CO
+        if (nextState.rasterParams.sampleShadingEnable && STATE_NEQ(rasterParams.minSampleShading)) {
+            COpenGLExtensionHandler::extGlMinSampleShading(nextState.rasterParams.minSampleShading);
+            UPDATE_STATE(rasterParams.minSampleShading);
+        }
+        if (STATE_NEQ(rasterParams.sampleMaskEnable)) {
+            disable_enable_fptr[nextState.rasterParams.sampleMaskEnable](GL_SAMPLE_MASK);
+            UPDATE_STATE(rasterParams.sampleMaskEnable);
+        }
+        if (nextState.rasterParams.sampleMaskEnable && STATE_NEQ(rasterParams.sampleMask[0])) {
+            COpenGLExtensionHandler::extGlSampleMaski(0u, nextState.rasterParams.sampleMask[0]);
+            UPDATE_STATE(rasterParams.sampleMask[0]);
+        }
+        if (nextState.rasterParams.sampleMaskEnable && STATE_NEQ(rasterParams.sampleMask[1])) {
+            COpenGLExtensionHandler::extGlSampleMaski(1u, nextState.rasterParams.sampleMask[1]);
+            UPDATE_STATE(rasterParams.sampleMask[1]);
+        }
+        if (STATE_NEQ(rasterParams.depthWriteEnable)) {
+            glDepthMask(nextState.rasterParams.depthWriteEnable);
+            UPDATE_STATE(rasterParams.depthWriteEnable);
+        }
+#undef STATE_NEQ
+#undef UPDATE_STATE
+    }
+    if (stateBits & GSB_VAO_AND_VERTEX_INPUT)
+    {
+    }
+    if (stateBits & GSB_DESCRIPTOR_SETS)
+    {
+        auto rebindBuffers = [this](GLenum target, const auto& currBufs, const auto& nextBufs)
+        {
+            auto bufBndDiffers = [](const auto& curr, const auto& next, uint32_t ix) {
+                return curr.buffers[ix] != next.buffers[ix] || curr.offsets[ix] != next.offsets[ix] || curr.sizes[ix] != next.sizes[ix];
+            };
+#define BUF_NEQ(ix) bufBndDiffers(currBufs, nextBufs, ix)
+
+            constexpr GLuint invalid_ix = static_cast<GLuint>(-1);
+            GLuint first = invalid_ix;
+            GLsizei count = 0u;
+            for (uint32_t i = 0u; i < 16u; ++i)
+            {
+                if (BUF_NEQ(i)) {
+                    if (first == invalid_ix)
+                        first = i;
+                    ++count;
+                }
+                else {
+                    if (count)
+                        COpenGLExtensionHandler::extGlBindBuffersRange(target, first, count, nextBufs.buffers+first, nextBufs.offsets+first, nextBufs.sizes+first);
+                    first = invalid_ix;
+                    count = 0u;
+                }
+            }
+#undef BUF_NEQ
+        };
+
+        //UBOs
+        rebindBuffers(GL_UNIFORM_BUFFER, currentState.descriptorsParams.ubos, nextState.descriptorsParams.ubos);
+        currentState.descriptorsParams.ubos = nextState.descriptorsParams.ubos;
+        //SSBOs
+        rebindBuffers(GL_SHADER_STORAGE_BUFFER, currentState.descriptorsParams.ssbos, nextState.descriptorsParams.ssbos);
+        currentState.descriptorsParams.ssbos = nextState.descriptorsParams.ssbos;
+
+        //textures
+        GLenum texTargets[80]{};
```

Comment:
same comment as for the buffer multibind

Reply:
this one is a little tricky, that's why im creating extra `texTagets` array here. For example: if previously in tex unit X was 2D texture and now we're going to bind 3D texture to tex unit X (or unbind anything - that is "bind" 0) i have to unbind texture 2D first and then bind the new one.

Reply:
ARB_multi_bind no longer gives a shit about the target.

Reply:
Only Intel Haswell on Windows 8 and below doesn't have this extension. You can add it to the required extension list.

Reply:
nice, that simplifies stuff

Reply:
you can actually kill the emulation code for that.